### PR TITLE
snap: ensure the state of snapshot flushed after applying snapshots.

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/17376

What's Changed:

This pr is an enhancement for the previous work https://github.com/tikv/tikv/pull/17408.

This pr is introduced to ensure that the state of region and snapshot will be flushed after applying
snapshot.

```commit-message
Ensure that the state of snapshot and region will be flushed into kvdb.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by overvenus — COMMENTED** [review]
Though it fixes the issue we discussed offline, I think we'd better to do fsync in `apply_sst_cf_file_without_ingest` as well, otherwise "correct" usage relies on caller does a fsync on WAL. Or add some comments to `apply_sst_cf_file_without_ingest` that explicitly tells caller that it requires callers to do a kv rocksdb WAL fsync.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/raftstore/src/store/snap/io.rs [hunk 1]
```diff
@@ -238,6 +238,9 @@ where
 
 /// Apply the given snapshot file into a column family. `callback` will be
 /// invoked after each batch of key value pairs written to db.
+///
+/// Attention, callers should manually flush the column family after applying
+/// all sst files to make sure the data is visible to the following operations.
```
Comment (overvenus) [inline components/raftstore/src/store/snap/io.rs:243]:
```suggestion
/// Attention, callers should manually flush and sync the column family after
/// applying all sst files to make sure the data durability.
```

**Review by SpadeA-Tang — APPROVED** [review]

File: components/raftstore/src/store/snap/io.rs [hunk 2]
```diff
@@ -363,6 +366,12 @@ where
     Ok(())
 }
 
+/// Apply the given snapshot file into a column family by directly writing kv
+/// pairs to db, without ingesting them. `callback` will be invoked after each
+/// batch of key value pairs written to db.
+///
+/// Attention, callers should manually flush the column family after applying
+/// all sst files to make sure the data is visible to the following operations.
```
Comment (overvenus) [inline components/raftstore/src/store/snap/io.rs:374]:
```suggestion
/// Attention, callers should manually flush and sync the column family after
/// applying all sst files to make sure the data durability.
```

**Pull request merged**